Fix for issue 6487: Opening BibTex file (doubleclick) from Folder with spaces not working

XDZhelheim opened the pull request:
Fixes #6487

- [x] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [x] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [x] Screenshots added in PR description (for UI changes)
- [x] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, submitted a pull request to the documentation repository.

Video:

https://user-images.githubusercontent.com/57553691/111628111-ce88ae00-882a-11eb-9dee-9dfe3087437f.mp4

Use `D:\TEST\test whitespace.bib` as an example.

Double-click it, and I got errors in console as the following:

```
Error opening file 'D:\TEST\test': Could not find a suitable import format.
Error opening file 'whitespace.bib': Could not find a suitable import format.
21:41:53.032 [JabRef - Remote Listener Server on port 6050] ERROR org.jabref.logic.importer.OpenDatabase - Error: File not found
```

Therefore, the path was regarded as two files: `D:\TEST\test` and `whitespace.bib`.

Start from `org.jabref.logic.importer.OpenDatabase` to backtrack the dataflow:

|         Class          |            Method            |      Variable      |
| :--------------------: | :--------------------------: | :----------------: |
|     `OpenDatabase`     |        `loadDatabase`        |    `fileToOpen`    |
|  `ArgumentProcessor`   |     `importAndOpenFiles`     |    `aLeftOver`     |
|      `JabRefCLI`       |         constructor          |       `args`       |
|  `ArgumentProcessor`   |         constructor          |       `args`       |
| `JabRefMessageHandler` | `handleCommandlineArguments` |     `message`      |
| `RemoteListenerServer` |       `handleMessage`        |     `argument`     |
| `RemoteListenerServer` |            `run`             | `input.getValue()` |
|       `Protocol`       |       `receiveMessage`       |     `argument`     |



Use logger to print the variable `argument`:

```java
if (argument instanceof String[]) {
    LOGGER.info(Arrays.toString((String[]) argument));
}
```

And I got:

```
21:41:53.024 [JabRef - Remote Listener Server on port 6050] INFO  org.jabref.logic.remote.shared.Protocol - [D:\TEST\test, whitespace.bib]
```

So I think the problem is that when passing the argument `D:\TEST\test whitespace.bib`, it will be considered as two arguments because there is a whitespace that serves as a delimiter.

I noticed that @zongxian said that a path containing Chinese characters would cause the same error. So the problem is not just related to delimiter, probably about text encoding.

Therefore I encoded the arguments in `RemoteClient.sendCommandLineArguments` and decoded them in `Protocol.receiveMessage`.

Then I created a binary and tested. Now it works well on my computer (Windows 10 1909).

tobiasdiez commented on src/main/java/org/jabref/logic/remote/client/RemoteClient.java:
```diff
@@ -52,7 +54,13 @@ public boolean ping() {
         try (Protocol protocol = openNewConnection()) {
-            protocol.sendMessage(RemoteMessage.SEND_COMMAND_LINE_ARGUMENTS, args);
+
+            String[] encodedArgs = args.clone();
```
Can you please put this addition in the `protocol.sendMessage` so that only the protocol has to worry about decoding/encoding. Also please add a short comment explaining why we need to encode. Thanks!

tobiasdiez reviewed (commented):
Thanks a lot for your detailed analysis!  The fix looks good to me.

The linked issue also mentions that double clicking a file without having jabref open doesn't work sometimes (if I got the gist right by skimming it). I guess this is not fixed with this PR, right?

XDZhelheim commented:
Thanks for your acknowledgement!

I moved it to `Protocol.sendMessage` and added some comments.

Besides, I created a new test case in `RemoteCommunicationTest`.

---

As @RickVandoorne mentioned:

> * If I double click on a .bib file while JabRef is NOT running then JabRef opens (it no longer gives me the error message). However, it does not open the file which I double clicked.
> * If I double click on a .bib file while JabRef is running it opens a new library called "untitled" but not the file I double clicked.
> * If I open JabRef first and then open a file using "File" → "Open library" it works fine.

I tested these three ways to open a `.bib` file, got no error.

Double click while JabRef not running:
https://user-images.githubusercontent.com/57553691/111728161-47c5e680-88a7-11eb-81d2-ddd56253b3da.mp4

Double click while JabRef running:
Video above.

"Open library":

https://user-images.githubusercontent.com/57553691/111728322-983d4400-88a7-11eb-983d-74ee3d45354c.mp4

XDZhelheim commented on src/test/java/org/jabref/logic/remote/RemoteCommunicationTest.java:
```diff
@@ -69,7 +69,17 @@ void commandLineArgumentTwoPassedToServer() {
+        client.sendCommandLineArguments(message.clone());
 
         verify(server).handleCommandLineArguments(message);
     }
```
I am wondering whether this test case is appropriate. Not so sure about how to write a test about socket communication.

tobiasdiez replied:
Yeah, it's hard to test this. I think, your added test is fine.

tobiasdiez commented on src/test/java/org/jabref/logic/remote/RemoteCommunicationTest.java:
```diff
@@ -51,7 +51,7 @@ void pingReturnsTrue() throws IOException, InterruptedException {
         final String[] message = new String[]{"my message"};
 
-        client.sendCommandLineArguments(message);
+        client.sendCommandLineArguments(message.clone());
```
I would move the clone call to the sendMessage method. It's usually a good practice to not modify parameters passed to a method.

XDZhelheim replied:
Done. Thank you for your advice!

Siedlerchr reviewed (approved):
thanks for the fix! LGTM as well

tobiasdiez reviewed (approved):
Thanks again for your contribution! Looking forward to your next PR 🚀
